src/main/java/servlets/UploadReceiptServlet.java

```diff
@@ -173,4 +164,65 @@ private static boolean isValidFilename(String filename) {
   private String getBlobServingUrl(BlobKey blobKey) {
     return "/serve-image?blob-key=" + blobKey.getKeyString();
   }
+
+  /**
+   * Extracts the raw text from the image with the Cloud Vision API. Returns a receipt
+   * entity populated with the extracted fields.
+   */
+  private Entity analyzeReceiptImage(String imageUrl, HttpServletRequest request)
+      throws ReceiptAnalysisException {
+    String absoluteUrl = getBaseUrl(request) + imageUrl;
+    AnalysisResults results = null;
+
+    try {
+      results = ReceiptAnalysis.serveImageText(absoluteUrl);
+    } catch (IOException e) {
+      throw new ReceiptAnalysisException("Receipt analysis failed.", e);
+    }
+
+    // TODO: Replace hard-coded values using receipt analysis with Cloud Vision.
+    double price = 5.89;
+    String store = "McDonald's";
+
+    // Create an entity with a kind of Receipt.
+    Entity receipt = new Entity("Receipt");
+    receipt.setProperty("price", price);
+    receipt.setProperty("store", store);
+    receipt.setUnindexedProperty("rawText", results.getRawText());
+
+    return receipt;
+  }
+
+  /**
+   * Get the base URL of the application for either the dev server or deployment.
+   */
+  private String getBaseUrl(HttpServletRequest request) {
+    String baseUrl = request.getScheme() + "://" + request.getServerName() + ":"
+        + request.getServerPort() + request.getContextPath();
+
+    // URL for dev server.
+    if (baseUrl.equals("http://0.0.0.0:80")) {
+      return "https://8080-9333ac09-1dd5-4f7f-8a7a-34f16c364c6b.us-east1.cloudshell.dev";
+    }
+
+    return baseUrl;
+  }
+
+  public class InvalidFileException extends Exception {
```

Comment:
Should we place these exceptions somewhere else? I might want to make use of some of them in ReceiptAnalysis, and it wouldn't make sense to import them from a servlet that depends on ReceiptAnalysis.

Reply:
I have moved the ReceiptAnalysisException into the shared class. I think the other two exceptions are specific to this class.